File: backend/src/services/reviewOrchestrator.ts
import { randomUUID } from "crypto";
import { ReviewRequest, ReviewResponse, Finding } from "../types.js";
import { runLLMReview } from "./llmClient.js";
import { scrubSecrets } from "../utils/scrubber.js";

const defaultConfig = {
  model: process.env.OPENAI_MODEL || "gpt-4o-mini",
  temperature: 0.2,
  maxTokens: 512
};

export async function orchestrateReview(request: ReviewRequest): Promise<ReviewResponse> {
  const started = performance.now();

  const scrubbedDiff = scrubSecrets(request.diff);
  const config = { ...defaultConfig, ...(request.config || {}) };

  const prompt = buildPrompt({
    diff: scrubbedDiff,
    repo: request.repo,
    prNumber: request.prNumber,
    filesChanged: request.filesChanged
  });

  const llmResult = await runLLMReview(prompt, config);
  const findings = normalizeFindings(llmResult.findings);

  const latencyMs = Math.round(performance.now() - started);

  return {
    findings,
    diagnostics: {
      provider: llmResult.provider,
      model: llmResult.model,
      latencyMs,
      tokens: llmResult.tokens,
      truncated: llmResult.truncated
    }
  };
}

function buildPrompt(input: { diff: string; repo?: string; prNumber?: string; filesChanged?: string[] }) {
  const header = [
    "You are an expert software reviewer.",
    "Given a git diff, identify issues and suggestions.",
    "Respond strictly as JSON array of findings with fields: severity, category, message, filePath, line, suggestion."
  ].join(" ");

  const meta = [
    input.repo ? `Repository: ${input.repo}` : null,
    input.prNumber ? `PR: ${input.prNumber}` : null,
    input.filesChanged?.length ? `Files changed: ${input.filesChanged.join(", ")}` : null
  ]
    .filter(Boolean)
    .join(" | ");

  return `${header}\n${meta}\nDiff:\n${input.diff}`;
}

function normalizeFindings(raw: Finding[]): Finding[] {
  return raw.map((item) => ({
    id: item.id || randomUUID(),
    severity: item.severity || "minor",
    category: item.category || "style",
    filePath: item.filePath,
    line: item.line,
    message: item.message || "Unspecified issue",
    suggestion: item.suggestion
  }));
}

File: backend_flask/review_orchestrator.py
import time
import uuid
from typing import Any, Dict, List

from llm_client import run_llm_review
from scrubber import scrub_secrets

DEFAULT_CONFIG = {
  "model": "gpt-4o-mini",
  "temperature": 0.2,
  "maxTokens": 512
}


def orchestrate_review(payload: Dict[str, Any]) -> Dict[str, Any]:
  started = time.time()

  diff = payload.get("diff")
  if not diff or not isinstance(diff, str):
    raise ValueError("diff is required")

  config = {**DEFAULT_CONFIG, **(payload.get("config") or {})}

  prompt = build_prompt({
    "diff": scrub_secrets(diff),
    "repo": payload.get("repo"),
    "prNumber": payload.get("prNumber"),
    "filesChanged": payload.get("filesChanged")
  })

  llm_result = run_llm_review(prompt, config)
  findings = normalize_findings(llm_result["findings"])

  return {
    "findings": findings,
    "diagnostics": {
      "provider": llm_result["provider"],
      "model": llm_result["model"],
      "latencyMs": int((time.time() - started) * 1000),
      "tokens": llm_result.get("tokens"),
      "truncated": llm_result.get("truncated", False)
    }
  }


def build_prompt(data: Dict[str, Any]) -> str:
  header = (
    "You are an expert software reviewer. "
    "Given a git diff, identify issues and suggestions. "
    "Respond strictly as JSON array of findings with fields: "
    "severity, category, message, filePath, line, suggestion."
  )

  meta = " | ".join(filter(None, [
    f"Repository: {data.get('repo')}" if data.get("repo") else None,
    f"PR: {data.get('prNumber')}" if data.get("prNumber") else None,
    f"Files changed: {', '.join(data.get('filesChanged') or [])}" if data.get("filesChanged") else None
  ]))

  return f"{header}\n{meta}\nDiff:\n{data['diff']}"


def normalize_findings(raw: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
  normalized = []
  for item in raw:
    normalized.append({
      "id": item.get("id") or str(uuid.uuid4()),
      "severity": item.get("severity", "minor"),
      "category": item.get("category", "style"),
      "filePath": item.get("filePath"),
      "line": item.get("line"),
      "message": item.get("message", "Unspecified issue"),
      "suggestion": item.get("suggestion")
    })
  return normalized
